clang/lib/Sema/SemaTemplateDeduction.cpp

```diff
@@ -5548,13 +5504,100 @@ static bool isAtLeastAsSpecializedAs(Sema &S,
+      // Otherwise the ordering rules for methods with expicit objet arguments
+      // against anything else make no sense.
+      shouldConvert1 = Method1 && !Method1->isExplicitObjectMemberFunction();
+      shouldConvert2 = Method2 && !Method2->isExplicitObjectMemberFunction();
+    } else {
+      // C++11 [temp.func.order]p3:
+      //   [...] If only one of the function templates is a non-static
+      //   member, that function template is considered to have a new
+      //   first parameter inserted in its function parameter list.
+      //
+      // Note that we interpret this to mean "if one of the function
+      // templates is a non-static member and the other is a non-member";
+      // otherwise, the ordering rules for static functions against non-static
+      // functions don't make any sense.
+      //
+      // C++98/03 doesn't have this provision but we've extended DR532 to cover
+      // it as wording was broken prior to it.
+      shouldConvert1 =
+          !Method2 && Method1 && Method1->isImplicitObjectMemberFunction();
+      shouldConvert2 =
+          !Method1 && Method2 && Method2->isImplicitObjectMemberFunction();
+    }
+    if (shouldConvert1) {
+      bool isR2 =
+          getLangOpts().CPlusPlus20 &&
+          (shouldConvert1
+               ? Method2->getRefQualifier() == RQ_RValue
+               : Proto2->param_type_begin()[0]->isRValueReferenceType());
+      // Compare 'this' from Method1 against first parameter from Method2.
+      ObjType1 = GetImplicitObjectParameterType(this->Context, Method1,
+                                                RawObjType1, isR2);
+      Args1.push_back(ObjType1);
+    }
+    if (shouldConvert2) {
+      bool isR1 =
+          getLangOpts().CPlusPlus20 &&
+          (shouldConvert2
+               ? Method1->getRefQualifier() == RQ_RValue
+               : Proto1->param_type_begin()[0]->isRValueReferenceType());
+      // Compare 'this' from Method2 against first parameter from Method1.
+      ObjType2 = GetImplicitObjectParameterType(this->Context, Method2,
+                                                RawObjType2, isR1);
+      Args2.push_back(ObjType2);
+    }
+    unsigned NumComparedArguments = NumCallArguments1 + shouldConvert1;
 
-  bool Better1 = isAtLeastAsSpecializedAs(*this, Loc, FT1, FT2, TPOC,
-                                          NumCallArguments1, Reversed);
-  bool Better2 = isAtLeastAsSpecializedAs(*this, Loc, FT2, FT1, TPOC,
-                                          NumCallArguments2, Reversed);
+    Args1.insert(Args1.end(), Proto1->param_type_begin(),
+                 Proto1->param_type_end());
+    Args2.insert(Args2.end(), Proto2->param_type_begin(),
+                 Proto2->param_type_end());
 
+    // C++ [temp.func.order]p5:
+    //   The presence of unused ellipsis and default arguments has no effect on
+    //   the partial ordering of function templates.
+    if (Args1.size() > NumComparedArguments)
```

Comment:
```suggestion
    Args1.resize(std::max(Args1.size(), NumComparedArguments));
    Args2.resize(std::max(Args2.size(), NumComparedArguments));
```

Reply:
`Args1.resize(std::max(Args1.size(), NumComparedArguments));` is not functionally equivalent to what I wrote, and will not work as intended. I need to drop elements, not to add new default constructed ones. I think you meant to write `min` instead of `max`. 
 Still I don't see why `Args1.resize(min(Args1.size(), NumComparedArguments));` is preferable to the original.

Reply:
Seemingly the original is more confusing as to what it is doing than I thought :)    I thought that the 'if' statements were overly verbose for something pretty trivial and the 1 liners made this less distracting along with reading the rest of the function.

Reply:
How can I make it less confusing? Should I put in a comment explaining that in `Args1` and `Args2` after the first `NumComparedArguments` elements there can only be unused ellipsis and default arguments? (It already states that  unused ellipsis and default arguments has no effect on the partial ordering of function templates. Just above the if)

Reply:
I thought the `resize` model was pretty clear, but that is a pattern I use frequently so I thought it read easily.

Reply:
Done